### PR TITLE
[feature] chandog-git/XP Meter 

### PR DESCRIPTION
**Pull request by chandog-git**
image

- initial implementation of xp meter
- xp per second, minute, hour
- xp gained per "tracked session"
- xp left to level 
- est time to level based on average xp

### CODE REVIEW

File: hordes-ui-mod.js [hunk 1]
```diff
@@ -382,941 +392,1071 @@
...
+        setGMChatVisibility: isGMChatVisible => {
+            const $chatGM = document.querySelector(`.js-chat-gm`);
+            state.chat.GM = isGMChatVisible;
+            $chatGM.classList.toggle('textgrey', !state.chat.GM);
+            modHelpers.filterAllChat();
+            save();
+        },
+
+        /** get char level */
+        getCurrentLvl: () => Number(document.querySelector('#ufplayer > div > div > div.progressBar.bgmana > .left').textContent.split('Lv. ')[1]),
```
Comment (timebomb0) [inline hordes-ui-mod.js]: If this method was named `getCurrentCharacterLvl`, we wouldn't even need the `get char level` comment above :D

File: hordes-ui-mod.js [hunk 2]
```diff
@@ -382,941 +392,1071 @@
...
+
+        /** get current xp */
+        getCurrentXp: () => Number(document.querySelector('#expbar > .bar > .progressBar > .left').textContent.split('/')[0].trim()),
+
+        /** get next level xp */
```
Comment (timebomb0) [inline hordes-ui-mod.js]: Hope you don't mind the nitpickiness here - I'm cool with remove comments above methods that type out the method name.
Your method names alone are good here :)
`getNextLevelXp` is just as descriptive as the comment `get next level xp`, so we don't necessarily need to repeat ourselves.

I'm personally a fan of using comments to explain the `why` of the code, assuming there's a reason or need to explain that.

Feel free to remove (or if you feel necessary, expand on) the comments above any other methods that you think apply here.

File: hordes-ui-mod.js [hunk 3]
```diff
@@ -382,941 +392,1071 @@
...
+            $chatGM.classList.toggle('textgrey', !state.chat.GM);
+            modHelpers.filterAllChat();
+            save();
+        },
+
+        /** get char level */
+        getCurrentLvl: () => Number(document.querySelector('#ufplayer > div > div > div.progressBar.bgmana > .left').textContent.split('Lv. ')[1]),
+
+        /** get current xp */
+        getCurrentXp: () => Number(document.querySelector('#expbar > .bar > .progressBar > .left').textContent.split('/')[0].trim()),
```
Comment (timebomb0) [inline hordes-ui-mod.js]: Is there a way to shrink the DOM selectors used here and elsewhere in this PR? Assuming they're still reliably the only single element that matches that selector.
Reason being: The structure `player > div > div > progressbar > left` is more likely to slightly change and stop working, as opposed to something like `.progressBar.bgmana .left`

File: hordes-ui-mod.js [hunk 4]
```diff
@@ -382,941 +392,1071 @@
...
+        /** user invoked reset of xp meter stats */
+        resetXpMeterState: () => {
+            state.xpMeterState.xpArr = [];
+            state.xpMeterState.averageXp = 0;
+            state.xpMeterState.gainedXp = 0;
+            document.querySelector('#timeremain').textContent = '-:-:-'
+        },
+
+        /** toggle the xp meter */
+        toggleXpMeter: () => {
```
Comment (timebomb0) [inline hordes-ui-mod.js]: Could we rename this method to `toggleXpMeterVisibility` and remove the above comment?

File: hordes-ui-mod.js [hunk 5]
```diff
@@ -382,941 +392,1071 @@
...
+            const xpMeterContainer = document.querySelector('#xpmeter')
+            xpMeterContainer.style.display === "none" ? xpMeterContainer.style.display = "block" : xpMeterContainer.style.display = "none";
+        },
+
+        msToString: (ms) => {
```
Comment (timebomb0) [inline hordes-ui-mod.js]: Could you move this function to `utils` near the bottom of the file?
It's not directly related to/reliant on any of the specific mods. I've been trying to keep these types of methods separate in utils, if that makes some sense.

File: hordes-ui-mod.js [hunk 6]
```diff
@@ -382,941 +392,1071 @@
...
-    			$name.classList.add('js-is-context-menu-initd');
-    			// Add name to element so we can target it in CSS when filtering chat for block list
-    			$name.setAttribute('data-chat-name', name);
-
-    			const showContextMenu = clickEvent => {
-    				// TODO: Is there a way to pass the name to showChatContextMenumethod, instead of storing in tempState?
-    				tempState.chatName = name;
-    				modHelpers.showChatContextMenu(name, {x: clickEvent.pageX, y: clickEvent.pageY});
-    			};
-    			$name.addEventL
```
Comment (timebomb0) [inline hordes-ui-mod.js]: Nitpicking here and elsewhere: Been using single quotes (apostrophes `'`) instead of double quotes `"` for strings in the code.
Setting DOM strings, like xpMeterHTMLString, with `"` in the HTML is totally fine though.

Comment (timebomb0) [inline hordes-ui-mod.js]: Can you remove all the svelte classes here? How painful would that be?

Trying really hard to avoid leveraging svelte classes for anything. They're a bit more unstable, more likely to change with updates. If you need styles from them, feel free to create `uidom-x` classes for custom styles, and copy the svelte CSS into there.

Comment (timebomb0) [inline hordes-ui-mod.js]: Any reason why there's backslashes at the end of every line?
I haven't had issues with multiline HTML strings that don't have these backslashes. Would it be ok to remove them?

Comment (timebomb0) [inline hordes-ui-mod.js]: Can remove unset attributes like this `style=""`

Comment (josh-channin) [reply hordes-ui-mod.js]: these are used for styling to apply the common UI theme. they are just the common classes used for the panels

Comment (josh-channin) [reply hordes-ui-mod.js]: done!

Comment (josh-channin) [reply hordes-ui-mod.js]: done!

Comment (josh-channin) [reply hordes-ui-mod.js]: done!

Comment (chandog-git) [reply hordes-ui-mod.js]: these are used by the main css bundle for the common theme UI.

File: hordes-ui-mod.js [hunk 7]
```diff
@@ -18,36 +18,45 @@
...
+        },
+        windowsPos: {},
+        blockList: {},
+        friendsList: {},
+        mapOpacity: 70, // e.g. 70 = opacity: 0.7
+        friendNotes: {},
+        chatTabs: [],
+        xpMeterState: {
+            currentXp: 0,
+            xpArr: [],
```
Comment (timebomb0) [inline hordes-ui-mod.js]: What do you tihnk about naming this something like `xpGains` and add a comment above this line, `Array of individual xp gains from mobs/etc`? Trying to improve ability to understand exactly what it is at a glance.

File: hordes-ui-mod.js (hunk 3 above)
Comment (timebomb0) [reply hordes-ui-mod.js]: Can this be `#expbar .progressBar > .left`?

File: hordes-ui-mod.js [hunk 8]
```diff
@@ -382,941 +391,1065 @@
...
+        // Then filters chat and saves updated chat state
+        setGMChatVisibility: isGMChatVisible => {
+            const $chatGM = document.querySelector(`.js-chat-gm`);
+            state.chat.GM = isGMChatVisible;
+            $chatGM.classList.toggle('textgrey', !state.chat.GM);
+            modHelpers.filterAllChat();
+            save();
+        },
+
+        getCurrentCharacterLvl: () => Number(document.querySelector('#ufplayer > div > div > div.progressBar.bgmana > .left').textContent.split('Lv. ')[1]),
```
Comment (timebomb0) [inline hordes-ui-mod.js]: Can this selector be `#ufplayer .bgmana > .left`?

File: hordes-ui-mod.js [hunk 9]
```diff
@@ -382,941 +391,1065 @@
...
+
+        getCurrentXp: () => Number(document.querySelector('#expbar > .bar > .progressBar > .left').textContent.split('/')[0].trim()),
+
+        getNextLevelXp: () => Number(document.querySelector('#expbar > .bar > .progressBar > .left').textContent.split('/')[1].trim()),
```
Comment (timebomb0) [inline hordes-ui-mod.js]: Can this selector be `#expBar .progressBar > .left`?

File: hordes-ui-mod.js [hunk 10]
```diff
@@ -382,941 +391,1065 @@
...
-    				modHelpers.showChatContextMenu(name, {x: clickEvent.pageX, y: clickEvent.pageY});
-    			};
-    			$name.addEventListener('click', showContextMenu); // Left click
-    			$name.addEventListener('contextmenu', showContextMenu); // Right click works too
-    		};
-    		Array.from(document.querySelectorAll('.name:not(.js-is-context-menu-initd)')).forEach(($name) => {
-    			addContextMenu($name, $name.textContent);
-    		});
-    		Array.from(document.querySelectorAll('.textwhisper .textf1:not(.js-is-context-menu-initd)')).forEach($whisperName => {
-    			// $w
```
Comment (timebomb0) [inline hordes-ui-mod.js]: Would prefer you use the preexisting `makeElement` method, for consistency's sake. Seen in utils near the bottom of the file, and usage examples throughout the file.

Comment (timebomb0) [inline hordes-ui-mod.js]: Can we make this a class, and prefix any class targeted by javascript with `js-`? So this may be a class named `js-sysxp` for example

Comment (timebomb0) [inline hordes-ui-mod.js]: Can you make this `id` a `js-xpmeter` class instead, like mentioned above?

Comment (timebomb0) [inline hordes-ui-mod.js]: Can you add a custom class, e.g. `js-xpmeter-icon`, to this `btn.black.svgicon` we define in the HTML above? Then we can just target that class here instead of having to dig into our `#xpmeter` DOM.

Comment (timebomb0) [inline hordes-ui-mod.js]: Can you add a custom class, e.g. `js-reset-button`, to this `div.btn` we define in the HTML above? Then we can just target that class here instead of having to dig into our `#xpmeter` DOM.

File: hordes-ui-mod.js [hunk 11]
```diff
@@ -18,36 +18,45 @@
...
     const tempState = {
-    	// The last name clicked in chat
-    	chatName: null,
-    	lastMapWidth: 0,
-    	lastMapHeight: 0,
+        // The last name clicked in chat
+        chatName: null,
+        lastMapWidth: 0,
+        lastMapHeight: 0,
+        xpMeterInterval: null,
```
Comment (timebomb0) [inline hordes-ui-mod.js]: Can you add a comment above this line, explaining briefly what `xpMeterInterval` is used for?

File: hordes-ui-mod.js (hunk 10 above)
Comment (chandog-git) [reply hordes-ui-mod.js]: Not sure what you mean here.

Comment (chandog-git) [reply hordes-ui-mod.js]: Not sure what you mean here.

Comment (chandog-git) [reply hordes-ui-mod.js]: I think you marked the wrong lines here.

Comment (chandog-git) [reply hordes-ui-mod.js]: I think you marked the wrong lines here.

Comment (chandog-git) [reply hordes-ui-mod.js]: I think you marked the wrong lines here.

Comment (chandog-git) [reply hordes-ui-mod.js]: I think you marked the wrong lines here.

Comment (chandog-git) [reply hordes-ui-mod.js]: I think you marked the wrong lines here.

File: hordes-ui-mod.js [hunk 12]
```diff
@@ -18,36 +18,45 @@
...
-(function() {
+(function () {
     'use strict';
 
     // If this version is different from the user's stored state,
     // e.g. they have upgraded the version of this script and there are breaking changes,
     // then their stored state will be deleted.
     const BREAKING_VERSION = 1;
-    const VERSION = '0.190'; // Should match version in UserScript description
+    const VERSION = '0.201'; // Should match version in UserScript description
```
Comment (timebomb0) [inline hordes-ui-mod.js]: 0.200 is fine, this version is for the public so going from .190 -> .200 makes sense.

File: hordes-ui-mod.js [hunk 13]
```diff
@@ -382,941 +391,1067 @@
...
+            state.xpMeterState.xpGains = []; /** array of xp deltas every second */
+            state.xpMeterState.averageXp = 0;
+            state.xpMeterState.gainedXp = 0;
+            document.querySelector('#timeremain').textContent = '-:-:-'
+        },
+
+        /** toggle the xp meter */
+        toggleXpMeterVisibility: () => {
+            const xpMeterContainer = document.querySelector('.js-xpmeter')
+            xpMeterContainer.style.display === "none" ? xpMeterContainer.style.display = "block" : xpMeterContainer.style.display = "none";
```
Comment (timebomb0) [inline hordes-ui-mod.js]: Nitpick: Use single quotes

File: hordes-ui-mod.js [hunk 14]
```diff
@@ -382,941 +391,1067 @@
...
+        getCurrentXp: () => Number(document.querySelector('#expbar .progressBar > .left').textContent.split('/')[0].trim()),
+
+        getNextLevelXp: () => Number(document.querySelector('#expbar .progressBar > .left').textContent.split('/')[1].trim()),
+
+        /** user invoked reset of xp meter stats */
+        resetXpMeterState: () => {
+            state.xpMeterState.xpGains = []; /** array of xp deltas every second */
+            state.xpMeterState.averageXp = 0;
+            state.xpMeterState.gainedXp = 0;
+            document.querySelector('#timeremain').textContent = '-:-:-'
```
Comment (timebomb0) [inline hordes-ui-mod.js]: missing semicolon

File: hordes-ui-mod.js (hunk 10 above)
Comment (timebomb0) [reply hordes-ui-mod.js]: Can we resolve this comment by using `makeElement`? Thanks!

File: hordes-ui-mod.js [hunk 15]
```diff
@@ -382,941 +391,1067 @@
...
-    				modHelpers.showChatContextMenu(name, {x: clickEvent.pageX, y: clickEvent.pageY});
-    			};
-    			$name.addEventListener('click', showContextMenu); // Left click
-    			$name.addEventListener('contextmenu', showContextMenu); // Right click works too
-    		};
-    		Array.from(document.querySelectorAll('.name:not(.js-is-context-menu-initd)')).forEach(($name) => {
-    			addContextMenu($name, $name.textContent);
-    		});
-    		Array.from(document.querySelectorAll('.textwhisper .textf1:not(.js-is-context-menu-initd)')).forEach($whisperName => {
-    			// $whis
```
Comment (timebomb0) [inline hordes-ui-mod.js]: As we talked about, could we replace all the `svelte-` classes here with `uidom-xpmeter-[name]` classes, replacing `[name]` with something appropriate. That way we wouldn't be relying on potentially fragile classes that may change.

Comment (timebomb0) [inline hordes-ui-mod.js]: If we created these `#xpm` etc classes in this mod, can we instead rename them to `.js-xpm` (classes not ids), for consistency's sake?

Comment (timebomb0) [inline hordes-ui-mod.js]: Feels like this function below would benefit from comments explaining briefly what it's doing.

Comment (timebomb0) [inline hordes-ui-mod.js]: For example, it'd be helpful to have a brief comment above this line explaining why we want to update the timeremaing if there is a positive averageXp

File: hordes-ui-mod.js (hunk 10 above)
Comment (chandog-git) [reply hordes-ui-mod.js]: `const $xpMeterElement = makeElement({ element: 'div', content: xpMeterHTMLString.trim() })
$layoutContainer.appendChild($xpMeterElement.firstChild);`

File: hordes-ui-mod.js [hunk 16]
```diff
@@ -382,941 +460,1071 @@
...
+            state.xpMeterState.xpGains = []; /** array of xp deltas every second */
+            state.xpMeterState.averageXp = 0;
+            state.xpMeterState.gainedXp = 0;
+            document.querySelector('.js-xp-time').textContent = '-:-:-';
+        },
+
+        /** toggle the xp meter */
+        toggleXpMeterVisibility: () => {
+            const xpMeterContainer = document.querySelector('.js-xpmeter');
+            xpMeterContainer.style.display === 'none' ? xpMeterContainer.style.display = 'block' : xpMeterContainer.style.display = 'none';
```
Comment (timebomb0) [inline hordes-ui-mod.js]: We can shorten this to `xpMeterContainer.style.display = xpMeterContainer.style.display === 'none' ? 'block' : 'none';`

File: hordes-ui-mod.js [hunk 17]
```diff
@@ -382,941 +460,1071 @@
...
-    				modHelpers.showChatContextMenu(name, {x: clickEvent.pageX, y: clickEvent.pageY});
-    			};
-    			$name.addEventListener('click', showContextMenu); // Left click
-    			$name.addEventListener('contextmenu', showContextMenu); // Right click works too
-    		};
-    		Array.from(document.querySelectorAll('.name:not(.js-is-context-menu-initd)')).forEach(($name) => {
-    			addContextMenu($name, $name.textContent);
-    		});
-    		Array.from(document.querySelectorAll('.textwhisper .textf1:not(.js-is-context-menu-initd)')).forEach($whisperName => {
-    			// $wh
```
Comment (timebomb0) [inline hordes-ui-mod.js]: Think we can remove the comment at the end of this line

Comment (timebomb0) [inline hordes-ui-mod.js]: Can you remove this commented console.log?

File: hordes-ui-mod.js [hunk 18]
```diff
@@ -382,941 +460,1071 @@
...
+            save();
+        },
+
+        getCurrentCharacterLvl: () => Number(document.querySelector('#ufplayer .bgmana > .left').textContent.split('Lv. ')[1]),
+
+        getCurrentXp: () => Number(document.querySelector('#expbar .progressBar > .left').textContent.split('/')[0].trim()),
+
+        getNextLevelXp: () => Number(document.querySelector('#expbar .progressBar > .left').textContent.split('/')[1].trim()),
+
+        /** user invoked reset of xp meter stats */
```
Comment (timebomb0) [inline hordes-ui-mod.js]: For consistency's sake, can we switch all the comments from `/**` to just `//`, and remove the ending `*/`?

File: hordes-ui-mod.js [hunk 19]
```diff
@@ -382,941 +460,1070 @@
...
-    				modHelpers.showChatContextMenu(name, {x: clickEvent.pageX, y: clickEvent.pageY});
-    			};
-    			$name.addEventListener('click', showContextMenu); // Left click
-    			$name.addEventListener('contextmenu', showContextMenu); // Right click works too
-    		};
-    		Array.from(document.querySelectorAll('.name:not(.js-is-context-menu-initd)')).forEach(($name) => {
-    			addContextMenu($name, $name.textContent);
-    		});
-    		Array.from(document.querySelectorAll('.textwhisper .textf1:not(.js-is-context-menu-initd)')).forEach($whisperName => {
-    			// $whisperName's textContent is "to [name]" or "fr
```
Comment (timebomb0) [inline hordes-ui-mod.js]: Can you use `makeElement` for `$xpMeterToggleElement` here, for cinsistency's sake?

Comment (timebomb0) [inline hordes-ui-mod.js]: We can invert this and keep it at the top of this function.
```
if (!document.querySelector('#expbar')) {
    return;
}

state.xpMeterState.....
```
That way we don't have to further nest all the code below.

File: hordes-ui-mod.js [hunk 20]
```diff
@@ -382,941 +460,1068 @@
...
+
+        getCurrentXp: () => Number(document.querySelector('#expbar .progressBar > .left').textContent.split('/')[0].trim()),
+
+        getNextLevelXp: () => Number(document.querySelector('#expbar .progressBar > .left').textContent.split('/')[1].trim()),
+
+        // user invoked reset of xp meter stats
+        resetXpMeterState: () => {
+            state.xpMeterState.xpGains = []; // array of xp deltas every second
+            state.xpMeterState.averageXp = 0;
+            state.xpMeterState.gainedXp = 0;
```
Comment (timebomb0) [inline hordes-ui-mod.js]: Can you save state after setting it here? `save()`

File: hordes-ui-mod.js [hunk 21]
```diff
@@ -382,941 +460,1068 @@
...
-    				modHelpers.showChatContextMenu(name, {x: clickEvent.pageX, y: clickEvent.pageY});
-    			};
-    			$name.addEventListener('click', showContextMenu); // Left click
-    			$name.addEventListener('contextmenu', showContextMenu); // Right click works too
-    		};
-    		Array.from(document.querySelectorAll('.name:not(.js-is-context-menu-initd)')).forEach(($name) => {
-    			addContextMenu($name, $name.textContent);
-    		});
-    		Array.from(document.querySelectorAll('.textwhisper .textf1:not(.js-is-context-menu-initd)')).forEach($whisperName => {
-    			// $whisperName's textContent is "to [name]" or "fr
```
Comment (timebomb0) [inline hordes-ui-mod.js]: Can you save state after setting it here? save()

**Comment by timebomb0** [comment]
Merged this in, thanks again -Updated the formatting and address some final feedback (saving state on set )in a separate branch.
